File: samples/java/DeliveryDispatch/Server/Configuration/src/main/java/systems/zlink/samples/deliverydispatch/server/configuration/DeliveryDispatchReadinessReporter.java
package systems.zlink.samples.deliverydispatch.server.configuration;

import org.springframework.context.SmartLifecycle;

import systems.zlink.framework.monitoring.ZLinkMeshPeerSnapshot;
import systems.zlink.framework.monitoring.ZLinkPeerState;
import systems.zlink.framework.monitoring.ZLinkRouteMeshRuntime;

import java.util.Arrays;
import java.util.List;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

/** Emits sample-owned readiness evidence from the public RouteMesh monitoring surface. */
public final class DeliveryDispatchReadinessReporter implements SmartLifecycle {
    private final ZLinkRouteMeshRuntime routeMeshRuntime;
    private final String nodeName;
    private final String meshName;
    private final List<String> actorRouteTargets;
    private boolean routeReported;
    private boolean running;
    private boolean[] actorRouteReported;
    private ScheduledExecutorService executor;

    public DeliveryDispatchReadinessReporter(
            ZLinkRouteMeshRuntime routeMeshRuntime,
            String nodeName,
            String meshName,
            String... actorRouteTargets) {
        this.routeMeshRuntime = routeMeshRuntime;
        this.nodeName = nodeName;
        this.meshName = meshName;
        this.actorRouteTargets = List.copyOf(Arrays.asList(actorRouteTargets));
        this.actorRouteReported = new boolean[actorRouteTargets.length];
    }

    @Override
    public synchronized void start() {
        if (running) {
            return;
        }
        running = true;
        executor =
                Executors.newSingleThreadScheduledExecutor(
                        runnable -> {
                            Thread thread = new Thread(runnable, "deliverydispatch-readiness");
                            thread.setDaemon(true);
                            return thread;
                        });
        executor.scheduleAtFixedRate(this::reportReadyState, 0, 100, TimeUnit.MILLISECONDS);
    }

    @Override
    public synchronized void stop() {
        running = false;
        if (executor != null) {
            executor.shutdownNow();
            executor = null;
        }
    }

    @Override
    public boolean isRunning() {
        return running;
    }

    @Override
    public boolean isAutoStartup() {
        return true;
    }

    @Override
    public int getPhase() {
        return 1;
    }

    private synchronized void reportReadyState() {
        if (!running || !routeMeshRuntime.isReady(meshName)) {
            return;
        }
        if (!routeReported) {
            routeReported = true;
            System.out.println("deliverydispatch-ready kind=route node=" + nodeName);
        }
        if (actorRouteTargets.isEmpty()) {
            return;
        }
        List<ZLinkMeshPeerSnapshot> peers = routeMeshRuntime.snapshot(meshName).peers();
        for (int index = 0; index < actorRouteTargets.size(); index++) {
            if (actorRouteReported[index] || !hasReadyPeer(peers, actorRouteTargets.get(index))) {
                continue;
            }
            actorRouteReported[index] = true;
            System.out.println(
                    "deliverydispatch-ready kind=actor-route node="
                            + nodeName
                            + " target="
                            + actorRouteTargets.get(index));
        }
    }

    private static boolean hasReadyPeer(List<ZLinkMeshPeerSnapshot> peers, String nodeName) {
        return peers.stream()
                .anyMatch(
                        peer ->
                                peer.nodeRid().toString().equals(nodeName)
                                        && peer.state() == ZLinkPeerState.READY);
    }
}

File: samples/kotlin/DeliveryDispatch/Server/Configuration/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/configuration/DeliveryDispatchReadinessReporter.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.configuration

import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit
import org.slf4j.LoggerFactory
import systems.zlink.framework.errors.ZLinkConfigurationException
import systems.zlink.framework.monitoring.ZLinkPeerState
import systems.zlink.framework.monitoring.ZLinkRouteMeshRuntime

/** Emits the sample-owned readiness evidence from the public, passive mesh view. */
class DeliveryDispatchReadinessReporter
private constructor(private val checks: List<ReadinessCheck>) : AutoCloseable {
    private val reporter: ScheduledExecutorService =
        Executors.newSingleThreadScheduledExecutor { runnable ->
            Thread(runnable, "deliverydispatch-readiness").apply { isDaemon = true }
        }

    private fun start(): DeliveryDispatchReadinessReporter {
        reporter.scheduleWithFixedDelay(::report, 0, 100, TimeUnit.MILLISECONDS)
        return this
    }

    private fun report() {
        checks.forEach(ReadinessCheck::reportIfReady)
        if (checks.all(ReadinessCheck::reported)) {
            reporter.shutdown()
        }
    }

    override fun close() {
        reporter.shutdownNow()
    }

    private class ReadinessCheck(private val evidence: String, private val ready: () -> Boolean) {
        var reported: Boolean = false
            private set

        fun reportIfReady() {
            if (reported) return
            try {
                if (ready()) {
                    logger.info(evidence)
                    reported = true
                }
            } catch (_: IllegalStateException) {
                // The public runtime view is available after Framework startup completes.
            } catch (_: ZLinkConfigurationException) {
                // The public runtime view is available after Framework startup completes.
            }
        }
    }

    companion object {
        private val logger = LoggerFactory.getLogger(DeliveryDispatchReadinessReporter::class.java)

        fun route(
            node: String,
            mesh: String,
            meshes: ZLinkRouteMeshRuntime,
        ): DeliveryDispatchReadinessReporter =
            DeliveryDispatchReadinessReporter(
                    listOf(
                        ReadinessCheck("deliverydispatch-ready kind=route node=$node") {
                            meshes.isReady(mesh)
                        }
                    )
                )
                .start()

        fun dispatch(meshes: ZLinkRouteMeshRuntime): DeliveryDispatchReadinessReporter =
            DeliveryDispatchReadinessReporter(
                    listOf(
                        ReadinessCheck(
                            "deliverydispatch-ready kind=route node=${SampleNames.DispatchNode}"
                        ) {
                            meshes.isReady(SampleNames.CourierSpotMesh)
                        },
                        ReadinessCheck(
                            "deliverydispatch-ready kind=actor-route node=${SampleNames.DispatchNode} " +
                                "target=${SampleNames.CourierNode1}"
                        ) {
                            hasReadyPeer(meshes, SampleNames.CourierNode1)
                        },
                        ReadinessCheck(
                            "deliverydispatch-ready kind=actor-route node=${SampleNames.DispatchNode} " +
                                "target=${SampleNames.CourierNode2}"
                        ) {
                            hasReadyPeer(meshes, SampleNames.CourierNode2)
                        },
                    )
                )
                .start()

        //  Confirm the named courier node, not a peer count. A count cannot tell which node is
        //  ready, so it passes on the wrong peer and fails when the count assumption drifts.
        private fun hasReadyPeer(meshes: ZLinkRouteMeshRuntime, target: String): Boolean {
            val courierMesh = meshes.snapshot(SampleNames.CourierSpotMesh)
            return courierMesh.peers.any {
                it.nodeRid().toString() == target && it.state() == ZLinkPeerState.READY
            }
        }
    }
}
